plat/renesas/rcar/aarch64/plat_helpers.S

```diff
@@ -0,0 +1,421 @@
+	 * plat_secondary_reset
+	 *
+	 * ---------------------------------------------
+	 */
+func plat_secondary_reset
+	mrs	x0, sctlr_el3
+	bic	x0, x0, #SCTLR_EE_BIT
+	msr	sctlr_el3, x0
+	isb
+
+	mrs	x0, cptr_el3
+	bic	w0, w0, #TCPAC_BIT
+	bic	w0, w0, #TTA_BIT
+	bic	w0, w0, #TFP_BIT
+	msr	cptr_el3, x0
+
+	mov_imm	x0, PARAMS_BASE
+	mov_imm	x2, BL31_BASE
+       ldr x3, =BOOT_KIND_BASE
+	mov x1, #0x1
+	str x1, [x3]
+	br	x2	/* jump to BL31 */
+	nop
+	nop
+	nop
+endfunc plat_secondary_reset
+
+	/* ---------------------------------------------
+	 * plat_enter_bl31
+	 *
+	 * ---------------------------------------------
+	 */
+func bl2_enter_bl31
+	mov	x20, x0
+        /*
+         * MMU needs to be disabled because both BL2 and BL31 execute
+         * in EL3, and therefore share the same address space.
+         * BL31 will initialize the address space according to its
+         * own requirement.
+         */
+#if RCAR_BL2_DCACHE == 1
+	/* Disable mmu and data cache */
+	bl	disable_mmu_el3
+	/* Data cache clean and invalidate */
+	mov	x0, #DCCISW
+	bl	dcsw_op_all
+	/* TLB invalidate all, EL3 */
+	tlbi	alle3
+#endif /* RCAR_BL2_DCACHE == 1 */
+	bl	disable_mmu_icache_el3
+	/* Invalidate instruction cache */
+	ic	iallu
+	dsb	sy
+	isb
+	ldp	x0, x1, [x20, #ENTRY_POINT_INFO_PC_OFFSET]
+	msr	elr_el3, x0
+	msr	spsr_el3, x1
+	eret
+endfunc bl2_enter_bl31
+
+	/* -----------------------------------------------------
+	 * void platform_mem_init (void);
+	 *
+	 * Zero out the mailbox registers in the shared memory
+	 * and set the rcar_boot_kind_flag.
+	 * The mmu is turned off right now and only the primary can
+	 * ever execute this code. Secondaries will read the
+	 * mailboxes using SO accesses.
+	 * -----------------------------------------------------
+	 */
+func platform_mem_init
+#if !IMAGE_BL2
+	ldr	x0, =MBOX_BASE
+	mov	w1, #PLATFORM_CORE_COUNT
+loop:
+	str	xzr, [x0], #CACHE_WRITEBACK_GRANULE
+	subs	w1, w1, #1
+	b.gt	loop
+#endif
+	ret
+endfunc platform_mem_init
+
+	/* ---------------------------------------------
+	 * void plat_report_exception(unsigned int type)
+	 * Function to report an unhandled exception
+	 * with platform-specific means.
+	 * ---------------------------------------------
+	 */
+func plat_report_exception
+#if IMAGE_BL2
+	mov	w1, #FIQ_SP_EL0
+	cmp	w0, w1
+	beq	rep_exec_fiq_elx
+	b	rep_exec_panic_type
+rep_exec_fiq_elx:
+	bl	gicv2_acknowledge_interrupt
+	mov	x2, #INT_ID_MASK
+	and	x0, x0, x2
+	mov	x1, #ARM_IRQ_SEC_WDT
+	cmp	x0, x1
+	bne	rep_exec_panic_id
+	mrs	x0, ELR_EL3
+	b	rcar_swdt_exec
+rep_exec_panic_type:
+	/* x0 is interrupt TYPE */
+	b	bl2_interrupt_error_type
+rep_exec_panic_id:
+	/* x0 is interrupt ID */
+	b	bl2_interrupt_error_id
+rep_exec_end:
+#endif
+	ret
+endfunc plat_report_exception
+
+	/* ---------------------------------------------
+	 * plat_is_my_cpu_primary
+	 *
+	 * ---------------------------------------------
+	 */
+func plat_is_my_cpu_primary
+	mrs	x0, mpidr_el1
+	and	x0, x0, #MPIDR_CPU_MASK
```

Comment:
As I've already noted there is a bug in this function which doesn't take in account MPIDR_CLUSTER_MASK and will cause a secondary CPU #0 from cluster #1 to be detected as a primary CPU.
The only reason why this code doesn't cause a problem when booting CPU #0 from cluster #1, is that it is never executed for any secondary CPU. Please see detailed explanation below.
In _bl31\aarch64\bl31_entrypoint.S_ macro _el3_entrypoint_common_ is called as:
	el3_entrypoint_common					\
		_init_sctlr=1					\
		_warm_boot_mailbox=!PROGRAMMABLE_RESET_ADDRESS	\
		_secondary_cold_boot=!COLD_BOOT_SINGLE_CPU	\
		_init_memory=1					\
		_init_c_runtime=1				\
		_exception_vectors=runtime_exceptions
with _warm_boot_mailbox=1 and secondary_cold_boot=1, because _plat\renesas\rcar\platform.mk_ defines
PROGRAMMABLE_RESET_ADDRESS	:= 0
COLD_BOOT_SINGLE_CPU		        := 0
RESET_TO_BL31			                := 1
This will expand _el3_entrypoint_common_ macro in _include\common\aarch64\el3_common_macros.S_ as
	.if \_warm_boot_mailbox
		/* -------------------------------------------------------------
		 * This code will be executed for both warm and cold resets.
		 * Now is the time to distinguish between the two.
		 * Query the platform entrypoint address and if it is not zero
		 * then it means it is a warm boot so jump to this address.
		 * -------------------------------------------------------------
		 */
		bl	plat_get_my_entrypoint
		cbz	x0, do_cold_boot
		br	x0

	do_cold_boot:
	.endif /* _warm_boot_mailbox */
Because code in _plat_secondary_reset()_ (_plat\renesas\rcar\aarch64\plat_helpers.S_)
       	ldr x3, =BOOT_KIND_BASE
	mov x1, #0x1
	str x1, [x3]
will always set warm reset type (RCAR_WARM_BOOT = 1), _plat_get_my_entrypoint()_ which checks boot type in X1
	/* warm boot or cold boot */
	and	x1, x21, #1
	cmp	x1, #0
	bne	warm_reset
will return X0 value != 0 read from the mail box:
warm_reset:
	/* --------------------------------------------------------------------
	 * A per-cpu mailbox is maintained in the trusted SDRAM. Its flushed out
	 * of the caches after every update using normal memory so its safe to
	 * read it here with SO attributes
	 * ---------------------------------------------------------------------
	 */
	ldr	x10, =MBOX_BASE
	bl	platform_get_core_pos
	lsl	x0, x0, #CACHE_WRITEBACK_SHIFT
	ldr	x0, [x10, x0]
	cbz	x0, _panic

In that case _do_cold_boot_ branch won't be taken: 
		bl	plat_get_my_entrypoint
		cbz	x0, do_cold_boot
		br	x0
do_cold_boot:

and code which checks for primary CPU:
		bl	plat_is_my_cpu_primary
		cbnz	w0, do_primary_cold_boot

won't be executed for any secondary.  Because secondary CPUs are always execute warm boot path and only single CPU comes from cold reset , COLD_BOOT_SINGLE_CPU should be set to 1. This option is also documented in docs\user-guide.rst:

-  **``COLD_BOOT_SINGLE_CPU``: This option indicates whether the platform may
   release several CPUs out of reset. It can take either 0 (several CPUs may be
   brought up) or 1 (only one CPU will ever be brought up during cold reset).
   Default is 0. If the platform always brings up a single CPU, there is no
   need to distinguish between primary and secondary CPUs and the boot path can
   be optimised. The ``plat_is_my_cpu_primary()`` and
   ``plat_secondary_cold_boot_setup()`` platform porting interfaces do not need
   to be implemented in this case."**

Reply:
But from what you are describing the warm path does just as expected - hence why I suppose there are no functional issues leaving system suspend, booting or hot-pluging cores. Is there a functional problem that we are trying to fix? 

Also do not forget that the cold path (ie, bl2) enters using exactly the same el3 macro and since only one core comes out of reset (from either cluster one or cluster two depending on a DIP switch setting), we need to be able to boot on any of those cases with the same TF-A; hence why we do not want to filter based on the cluster.

So yeah, the actual _functionality_ seems correct to me. Also worth noticing that if this was indeed a functional bug Renesas would have had many issues reported on their many platforms since this implementation is a 1:1 map of theirs (see [1])

As we discussed, I can of course see that there is a real semantic bug, that is for sure...I mean that function would identify as primary the same cpu on either core; however, for this platform, this seems to be intended.

[1] https://github.com/renesas-rcar/arm-trusted-firmware/blob/rcar_gen3/plat/renesas/rcar/aarch64/rcar_helpers.S#L220

Reply:
and yes setting COLD_BOOT_SINGLE_CPU removes  the semantic issue and therefore the unnecessary check.  Much better like this. cool.

Reply:
tests are ok. will push an update now